### PR TITLE
Restore start_tests() for result plugins [v3]

### PR DESCRIPTION
**Pull request by apahim**
With commit 2e6ecce6, the start_tests() method from results was lost.
But the runner was using that method to report the number of tests to
the UI.

This patch re-creates the start_tests() method, now using the new plugin
interface, fixing the tests number information in the UI.

Reference: https://trello.com/c/62I3ABZ0
Signed-off-by: Amador Pahim <apahim@redhat.com>

---
v3:
- Cover `remote` runner.
- Style fix.

v2: #1768 
- Fix the same issue in TAP plugin as well.

v1: #1766

### CODE REVIEW

**Comment by clebergnu** [comment]
@apahim this is not a review, but a comment to raise a discussion.  First of all, I tested this code and it works as intended, that is, it fixes the issue.

But, I'm skeptical. I wonder if adding another job phase (and hook) that refers to the "start of tests" is the way to go.

I'm stating that mainly because I realized that, depending on how you look at it, the contents of `test_suite` is indeed the number of tests we intend to run.  We're now calling the mechanism that creates the variations of the executed the *tests* the varianter, which is *not* going to be optional anytime soon (individual varianter plugin implementations can be).  I wonder if we should present the number of variants, something like:

```
$ ./scripts/avocado run sleeptest.py -m examples/tests/sleeptest.py.data/sleeptest.yaml 
JOB ID     : b4acc4a4684c36b1c38a82ebc7833328b8ec1a78
JOB LOG    : /home/cleber/avocado/job-results/job-2017-02-09T21.54-b4acc4a/job.log
TESTS      : 1
VARIANTS: 4
TEST VARIANTS: 4 
 (1/4) sleeptest.py:SleepTest.test;1: PASS (1.01 s)
 (2/4) sleeptest.py:SleepTest.test;2: PASS (1.01 s)
 (3/4) sleeptest.py:SleepTest.test;3: PASS (1.01 s)
 (4/4) sleeptest.py:SleepTest.test;4: PASS (1.01 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
TESTS TIME : 4.04 s
JOB HTML   : /home/cleber/avocado/job-results/job-2017-02-09T21.54-b4acc4a/html/results.html
```

Or something like:

```
$ ./scripts/avocado run sleeptest.py -m examples/tests/sleeptest.py.data/sleeptest.yaml 
JOB ID     : b4acc4a4684c36b1c38a82ebc7833328b8ec1a78
JOB LOG    : /home/cleber/avocado/job-results/job-2017-02-09T21.54-b4acc4a/job.log
TESTS      : 4 (1 BASE TEST, 4 VARIANTS)
 (1/4) sleeptest.py:SleepTest.test;1: PASS (1.01 s)
 (2/4) sleeptest.py:SleepTest.test;2: PASS (1.01 s)
 (3/4) sleeptest.py:SleepTest.test;3: PASS (1.01 s)
 (4/4) sleeptest.py:SleepTest.test;4: PASS (1.01 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
TESTS TIME : 4.04 s
JOB HTML   : /home/cleber/avocado/job-results/job-2017-02-09T21.54-b4acc4a/html/results.html
```

Those would expose the varianter to the end user.

One other approach that came to me is to refactor the varianter and make it avaible earlier to the job, before `Runner.run_suite()` (IIRC).

**Comment by apahim** [comment]
@clebergnu ok, it makes sense not to create yet another job phase. To fix this regression, I'd go with the refactor of the variant initialization so the job can have the number of tests earlier and update the result.tests_total before calling `run_tests()` as you suggested.
If there's a request to separate the number of tests from the number of variants, then we can think in how to organize it in our UI, but I don't see it as useful right now.
I'm working in the refactor. Thank you.

**Comment by apahim** [comment]
Ok, let me rephrase this. The problem with making the `variant` available earlier in the job and let job calculate the total number of tests is when we are using a remote plugin. In that case, we don't even have the test file locally to calculate the number of tests. That was the main reason I moved the number of tests calculation to the runner, so a remote runner can provide that information to be exhibited in the UI.
Now, question is: what is the best way to let the runner provide this information to the UI? Since the runner is engaged after the pre_tests hook, we need either a new phase, as proposed in this PR, or at least, a result event that can be triggered from the runner to write in the output.

**Comment by ldoktor** [comment]
Or we could rely on the `start_test` method as shown in the `TAP` fix https://github.com/avocado-framework/avocado/pull/1779 which works even for dynamically calculated number of tests (once the runner knows how many tests there will be it can propagate/update this number.
